File: LeetCode/Problems/Python/403. Frog Jump.py
#Dynamic programming
#Time: O(n^2)
#Space: O(n^2)
class Solution:
    def canCross(self, stones: List[int]) -> bool:
        if not stones:
            return False
        S = len(stones)
        
        dp = dict((x,set()) for x in stones)
        
        if 1 in dp:
            dp[1].add(1)
        else:
            return False
        
        for i in range(1, S):
            for k in dp[stones[i]]: 
                for step in [k-1,k,k+1]:
                    if step > 0 and stones[i]+step in dp:
                        dp[stones[i]+step].add(step)

        return dp[stones[-1]] != set()

File: LeetCode/Problems/C++/403. Frog Jump.cpp
class Solution {
public:
    bool canCross(vector<int>& stones) {
        if(!stones.size()) return false;
        
        unordered_map<int, unordered_set<int>> dp;
        for(int i:stones){
            dp[i] = {};
        }
        
        if(dp.find(1)!=dp.end())
            dp[1].insert(1);
        else
            return false;
                
        for(int i:stones){
            for(int k:dp[i]){
                for(int step=k-1;step<k+2;step++){
                    if(step>0 && dp.find(i+step)!=dp.end()){
                        dp[i+step].insert(step);
                        if(i+step == stones.back()) return true;
                    }
                }
            }
        }
        return !dp[stones.back()].empty();
    }
};